File: problems/python/ParkingLotSystem/src/service/payment_service.py
"""Delegates payment to a configured strategy (used at exit)."""
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from model.parking_ticket import ParkingTicket
    from strategy.payment.payment_strategy import PaymentStrategy


class PaymentService:
    def __init__(self, strategy: "PaymentStrategy") -> None:
        self._strategy = strategy

    def pay(self, ticket: "ParkingTicket", fee: float) -> bool:
        success = self._strategy.pay(fee, ticket)
        if not success:
            print(f"Payment failed for ticket: {ticket.get_ticket_id()}")
        return success

File: problems/java/ParkingLotSystem/src/service/PaymentService.java
package service;

import model.ParkingTicket;
import strategy.payment.PaymentStrategy;

/**
 * Delegates payment to a configured strategy. Used at exit before releasing the spot.
 */
public class PaymentService {
    private final PaymentStrategy strategy;

    public PaymentService(PaymentStrategy strategy) {
        this.strategy = strategy;
    }

    /**
     * Processes the payment for the given ticket and fee.
     *
     * @param ticket the parking ticket
     * @param fee    the amount to charge
     * @return true if payment succeeded, false otherwise
     */
    public boolean pay(ParkingTicket ticket, double fee) {
        boolean success = strategy.pay(fee, ticket);
        if (!success) {
            System.out.println("Payment failed for ticket: " + ticket.getTicketId());
        }
        return success;
    }
}
